types.d.ts

```diff
@@ -0,0 +1,699 @@
+import { EventEmitter } from "events";
+import https from 'https';
+import * as core from 'express-serve-static-core'
+
+//TODO: need to better understand this type
+type InputType = {}
+
+
+export interface Subscriber {
+  send(message: string): void
+}
+
+export interface ActionTypeClass {
+  new(thing: Thing, input: InputType): ActionTypeClass
+}
+
+/**
+ * An Action represents an individual action on a thing.
+ */
+export declare class Action {
+
+  constructor(id: string, thing: Thing, name: string, input: InputType);
+
+  /**
+   * Get the action description.
+   *
+   * @returns {Object} Description of the action as an object.
+   */
+  asActionDescription(): Action.ActionDescription;
+
+  /**
+   * Set the prefix of any hrefs associated with this action.
+   *
+   * @param {String} prefix The prefix
+   */
+  setHrefPrefix(prefix: string);
+
+  /**
+   * Get this action's ID.
+   *
+   * @returns {String} The ID.
+   */
+  getId(): string
+
+  /**
+   * Get this action's name.
+   *
+   * @returns {String} The name.
+   */
+  getName(): string
+
+  /**
+   * Get this action's href.
+   *
+   * @returns {String} The href.
+   */
+  getHref(): string
+
+  /**
+   * Get this action's status.
+   *
+   * @returns {String} The status.
+   */
+  getStatus(): string;
+
+  /**
+   * Get the thing associated with this action.
+   *
+   * @returns {Object} The thing.
+   */
+  getThing(): Thing
+
+  /**
+   * Get the time the action was requested.
+   *
+   * @returns {String} The time.
+   */
+  getTimeRequested(): string
+  /**
+   * Get the time the action was completed.
+   *
+   * @returns {String} The time.
+   */
+  getTimeCompleted(): string
+  /**
+   * Get the inputs for this action.
+   *
+   * @returns {Object} The inputs.
+   */
+  getInput(): InputType
+
+  /**
+   * Start performing the action.
+   */
+  start(): void;
+
+  /**
+   * Override this with the code necessary to perform the action.
+   *
+   * @returns {Object} Promise that resolves when the action is finished.
+   */
+  performAction(): Promise<any>;
+
+  /**
+   * Override this with the code necessary to cancel the action.
+   *
+   * @returns {Object} Promise that resolves when the action is cancelled.
+   */
+  cancel(): Promise<any>;
+
+  /**
+   * Finish performing the action.
+   */
+  finish(): void
+}
+
+declare namespace Action {
+  //TODO need to better understand this type
+  type ActionMetadata = any
```

Comment:
It might help to look at https://iot.mozilla.org/wot/, especially [this section](https://iot.mozilla.org/wot/#action-object) for more information about the less documented shapes

Reply:
Will do!

Reply:
This has been improved in https://github.com/mozilla-iot/webthing-node/pull/91/commits/c827b37b064e867e4db74971feaa6b2069b4dcb1